pkg/spanconfig/spanconfigsqlwatcher/buffer.go

```diff
@@ -119,27 +185,94 @@ func (b *buffer) flushEvents(
-			return ei.timestamp.Less(ej.timestamp)
-		}
-		return ei.update.ID < ej.update.ID
+	bufferedEvents := make(bufferedEvents, 0, len(events))
+	for _, ev := range events {
+		bufferedEvents = append(bufferedEvents, ev.(event))
+	}
+	// Nil out the underlying slice since we have copied over the events.
+	events = nil
+
+	// The events slice can contain a SQLUpdate that applies to a descriptor, or a
+	// SQLUpdate that applies to a protected timestamp record. We sort the
+	// protected timestamp SQLUpdates to the end of the slice, so that we can
+	// perform the subsequent deduplication of SQLUpdates. We do this instead of
+	// pre-processing events to separate the two kinds of updates, to save on the
+	// allocation of an additional `events` slices.
+	sort.Sort(bufferedEvents)
+
+	// Find the index before which all events are SQLUpdates on descriptors.
+	descriptorUpdatesIdx := sort.Search(len(bufferedEvents), func(i int) bool {
+		update := bufferedEvents[i].update
+		return update.IsProtectedTimestampUpdate()
 	})
-	for i, ev := range events {
-		if i == 0 || events[i-1].(event).update.ID != ev.(event).update.ID {
-			updates = append(updates, ev.(event).update)
+
+	// Deduplicate the SQLUpdates emitted by the buffer that apply to descriptors.
+	for i, ev := range bufferedEvents[:descriptorUpdatesIdx] {
+		update := ev.update
+		descriptorUpdate := update.GetDescriptorUpdate()
+		if i == 0 {
+			sqlUpdates = append(sqlUpdates, update)
 			continue
 		}
-		descType, err := combine(updates[len(updates)-1].DescriptorType, ev.(event).update.DescriptorType)
+
+		prevUpdate := bufferedEvents[i-1].update
+		if prevUpdate.GetDescriptorUpdate().ID != descriptorUpdate.ID {
+			sqlUpdates = append(sqlUpdates, update)
+			continue
+		}
+
+		prevDescriptorSQLUpdate := sqlUpdates[len(sqlUpdates)-1].GetDescriptorUpdate()
+		descType, err := combine(prevDescriptorSQLUpdate.Type,
+			descriptorUpdate.Type)
 		if err != nil {
 			return nil, hlc.Timestamp{}, err
 		}
-		updates[len(updates)-1].DescriptorType = descType
+		sqlUpdates[len(sqlUpdates)-1] = spanconfig.MakeDescriptorSQLUpdate(
+			prevDescriptorSQLUpdate.ID, descType)
 	}
-	return updates, combinedFrontierTS, nil
+
+	// Truncate the slice to only include ProtectedTimestampUpdates now that we
+	// have copied over all de-duped DescriptorUpdates.
+	bufferedEvents = bufferedEvents[descriptorUpdatesIdx:]
+
+	// Deduplicate the SQLUpdates emitted by the buffer that apply to protected
+	// timestamps.
+	for i, ev := range bufferedEvents {
+		update := ev.update
+		curUpdate := update.GetProtectedTimestampUpdate()
+		if i == 0 {
+			sqlUpdates = append(sqlUpdates, update)
+			continue
+		}
+
+		prevUpdate := bufferedEvents[i-1].update.GetProtectedTimestampUpdate()
+
+		// If the previous buffered event, and the current event are different
+		// target ProtectedTimestampUpdates there is no deduplication possible.
+		if prevUpdate.IsClusterUpdate() && curUpdate.IsTenantsUpdate() ||
```

Comment:
[nit] Can this be simplified?
```go
if prevUpdate.IsClusterUpdate() != curUpdate.IsClusterUpdate() {
```

Reply:
Alternatively, if this case was at the last thing we were checking, we could just directly append to the slice.

```go
// If the previous buffered event, and the current event are both cluster
// target ProtectedTimestampUpdates we can ignore duplicates.
if prevUpdate.IsClusterUpdate() && curUpdate.IsClusterUpdate() {
  continue
}

// If the previous buffered event, and the current event are both tenant
// target ProtectedTimestampUpdates, compare tenantIDs and ignore
// duplicates.
if prevUpdate.IsTenantsUpdate() && curUpdate.IsTenantsUpdate() {
  if prevUpdate.TenantTarget != curUpdate.TenantTarget {
    sqlUpdates = append(sqlUpdates, update)
    continue
  }
}

// The previous buffered event and the current event are different
// target ProtectedTimestampUpdates, there is no deduplication possible.
sqlUpdates = append(sqlUpdates, update)
```